sphereMove/SphereMove.cs

```diff
@@ -152,56 +200,115 @@ IEnumerator DelayedUpdate()
-                    car.date = interpolatedPosTime[i];
-                    Vector3 newPosition = new Vector3(car.x, car.y, this.gameObject.transform.localPosition.z);
-                    this.gameObject.transform.localPosition = newPosition;
-                    Debug.Log("Update x=" + this.gameObject.transform.localPosition.x.ToString() + "y=" + this.gameObject.transform.localPosition.y.ToString());
-
-                    // Wait until the next time to update the position
-                    yield return new WaitForSeconds((float)(interpolatedPosTime[i + 1] - interpolatedPosTime[i]).TotalSeconds);
-
-                }
-
-                // All interpolated positions have been used. Calculate new trajectory in the other coroutine
-                new_trajectory = true;
+                case State.Begin:
+                    if (!need_new_trajectory)
+                    {
+                        // TODO, if at the beginning stage the car is still, will give a NaN traj and give an error, ut should be fine bcs it changes to the Running stage quickly
+                        current_tracking_time = listTime[0];
+                        for (int i = 0; i < interpolatedPosX.Count; i++)
+                        {
+                            Thread.Sleep((int) Math.Round((interpolatedPosTime[i] - current_tracking_time).TotalSeconds * 1000)); // Blocked sleep in ms
+                            car.x = interpolatedPosX[i];
+                            car.y = interpolatedPosY[i];
+                            car.date = interpolatedPosTime[i].ToString("yyyy-MM-ddTHH:mm:ss.ffffff");
+                            Vector3 newPosition = new Vector3(car.x, car.y, this.gameObject.transform.localPosition.z);
+                            this.gameObject.transform.localPosition = newPosition;
+                            Debug.Log("Update x=" + this.gameObject.transform.localPosition.x.ToString() + "y=" + this.gameObject.transform.localPosition.y.ToString());
+                            current_tracking_time = interpolatedPosTime[i];
+                        }
+                        interpolatedPosTime.Clear();
+                        interpolatedPosX.Clear();
+                        interpolatedPosY.Clear();
+                        need_new_trajectory = true;
+                        currentState = State.Running;
+                        Debug.Log("State -> Running");
+                    }
+                    yield return new WaitForSeconds((float)0.0001);
+                    break;
+                case State.Running:
+                    if (!need_new_trajectory)
+                    {
+                        for (int i = 0; i < interpolatedPosX.Count; i++)
+                        {
+                            // TODO: if we let it sleep until the next position point, it will anyway be unsmooth
+                            Thread.Sleep((int)Math.Round((interpolatedPosTime[i] - current_tracking_time).TotalSeconds * 1000)); // Blocked sleep in ms

```

Comment:
Why do we thread block here of we have the `WaitForSeconds` at the bottom? Is this causing the jittering motion? What if we reduced this?

Reply:
Yes it is causing the jittering motion but it's just the algorithm we designed. We set the position instead of speed, so it will never be smooth unless the position update in both distance interval and time interval is not obvious with naked eyes. If you still remember, what we wanted to achieve is, if traj[0] and traj[1] have time interval 1ms, we want the update to happen exactly 1ms later. So my intuition here is to always block the thread until the current trajectory is done. About the synchronization of the two coroutines, I would say there exists better practice like using yield return WaitUntil(need_new_trajectory) in GetCarData instead of letting it grab any priority (not sure) after WaitForSeconds(Time). But for the jittering motion, I suppose it is inevitable in the current algorithm, especially when the car is fast.

Reply:
Hmm so what if we instead of setting `this.gameObject.transform.localPosition` we set the localSpeed and interpolate the speed for the vehicle (I think we would only need the speed at current position and the speed at the next position). We keep our position interpolation as well but also add in the speed interpolation if that is possible. I feel like this has more to do with the way unity handles the speed and positioning of an object vs how we coded it.

Reply:
> Hmm so what if we instead of setting `this.gameObject.transform.localPosition` we set the localSpeed and interpolate the speed for the vehicle (I think we would only need the speed at current position and the speed at the next position). We keep our position interpolation as well but also add in the speed interpolation if that is possible. I feel like this has more to do with the way unity handles the speed and positioning of an object vs how we coded it.

Well, how unity handles it depends on how we code it. If we want to use both speed and position, we ourselves need to figure out how to synchronize the two. The simplest way is, calculate the average velocity and orientation between trajectory points, let some fake_velocity=displacement/time and let the car run with set velocity nstead of set any position. This intuitively will give a continuous trajectory in position, like connecting all trajectory points using straight lines. For sure, it is not continuous in velocity and you can imagine the straight-line connected trajectory. By "speed interpolating", I'm quite confused. I personally do not want to fetch velocity from openF1 because, 1. I did not check, whether the position and speed magnitude data are synchronized with the same timestamp. They need to be fetched from two separate urls; 2. If they are somehow luckily synchronized in time, we need to write a new Cubic function to interpolate position and with each points' velocity. Then this Cubic is gonna return you a trajectory of points, then what, it comes back to the same problem that w

Reply:
we need to decide what velocity to set for the car.

With out acceleration/force info, it's never a real physical phenomenon. I think what we can do is to make it as real as possible, if we 1. schedule the coroutines' execution correctly; 2. the straight-line connected trajectory is still not satisfiable; 3. we can ask for more trajectory points when the car is fast (instead of setting M=35 globally).